Fixed findAll with order on custom field, limit, and hasMany relations

troyastorino opened the pull request:
Fixed #5332

troyastorino commented:
I looked at the failing tests, and am skeptical that this change would have broken those tests. Would a rebuild on the [mysql build](https://travis-ci.org/sequelize/sequelize/jobs/105596675) be worth trying? And I don't really know where to start looking on the [postgres-native build](https://travis-ci.org/sequelize/sequelize/jobs/105596666).

Any help would be appreciated, thanks!

mickhansen commented:
We have some intermittent failures, so likely not related to your changes (also green now).

mickhansen commented:
Do you mind adding a unit test for the sql generation aswell? Here: https://github.com/sequelize/sequelize/blob/master/test/unit/sql/select.test.js or create a new file for order specific tests

troyastorino commented:
OK, I added some query generation tests, but two of the tests I added are failing for MSSQL. I tried fiddling around with them to make them work, but I'm not familiar with MSSQL, so I don't know if the tests are actually showing a problem, or (much more likely) I'm just not sure how they should be working.

mickhansen commented:
Linting errors, all builds are failing so i can't look at what might be failing with MSSQL.

troyastorino commented:
Whoops, sorry! OK, fixed the linting, and now only mssql is failing

troyastorino commented:
Basically, I'm not sure whether this query (https://travis-ci.org/sequelize/sequelize/jobs/106788884#L1832):
```sql
SELECT [User].*, [Posts].[id] AS [Posts.id], [Posts].[le_title] AS [Posts.leTitle]
FROM (
  SELECT [User].[first_name] AS [firstName], [User].[id] AS [id] FROM [users] AS [User] 
  ORDER BY [User].[first_name] DESC, [id] OFFSET 0 ROWS FETCH NEXT 100 ROWS ONLY
) AS [User] 
LEFT OUTER JOIN [posts] AS [Posts]
ON [User].[id] = [Posts].[user_id] ORDER BY [User].[firstName] DESC;
```
is supposed to have `, [id]` inside of it

mickhansen commented:
Hmmm, mssql is abit odd, we might be added the primary key by default when doing limit.
https://github.com/sequelize/sequelize/blob/master/lib/dialects/mssql/query-generator.js#L622

troyastorino commented:
Ya, I was looking at that, but that's what was getting me confused (the conditions around that line). If that's proper behavior, I can fiddle around with generating the expected query to make sure it matches.

troyastorino commented:
But, since the integration tests worked, I'd assume that it's the right behavior, right?

mickhansen commented:
Yeah i guess that's the expected behaviour :)
